presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,203 @@ private boolean advanceToNextRowGroup()
+        if (fieldDefinitionLevels != null) {
+            int maxDefinitionLevel = getMaxDefinitionLevel(path);
+            int maxRepetitionLevel = getMaxRepetitionLevel(path);
+            for (int i = 0; i < fieldDefinitionLevels.size(); i++) {
+                if (fieldRepetitionLevels.get(i) <= maxRepetitionLevel) {
+                    if (isNullField(isRequired(path), fieldDefinitionLevels.get(i), maxDefinitionLevel)) {
+                        structIsNull.add(true);
+                        structOffsets.add(offset);
+                    }
+                    else if (isStructDefined(fieldDefinitionLevels, maxDefinitionLevel, i)) {
+                        structIsNull.add(false);
+                        offset++;
+                        structOffsets.add(offset);
+                    }
+                }
+            }
+            definitionLevels.addAll(fieldDefinitionLevels);
+            repetitionLevels.addAll(fieldRepetitionLevels);
         }
+    }
+
+    private boolean isStructDefined(IntList fieldDefinitionLevels, int maxDefinitionLevel, int i)
+    {
+        return fieldDefinitionLevels.get(i) >= maxDefinitionLevel;
+    }
 
-        int blockSize = blocks[0].getPositionCount();
-        int[] offsets = new int[blockSize + 1];
-        for (int i = 1; i < offsets.length; i++) {
-            elementOffsets.add(parameters.size());
-            offsets[i] = i;
+    /**
+     * Each collection (Array or Map) has four variants of presence:
+     * 1) Collection is not defined, because one of it's optional parent fields is null
+     * 2) Collection is null
+     * 3) Collection is defined but empty
+     * 4) Collection is defined and not empty. In this case offset value is increased by the number of elements in a collection.
+     */
+    private void calculateCollectionsOffsets(String[] path, IntList offsets, BooleanList collectionIsNull, IntList definitionLevels, IntList repetitionLevels)
+    {
+        int maxDefinitionLevel = getMaxDefinitionLevel(path);
+        int maxRepetitionLevelOfCollectionElement = getMaxRepetitionLevel(path) + 1;
+        int offset = 0;
+        offsets.add(offset);
+        for (int i = 0; i < definitionLevels.size(); i = getNextCollectionIndex(repetitionLevels, maxRepetitionLevelOfCollectionElement, i)) {
+            if (isNullField(isRequired(path), definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(true);
+                offsets.add(offset);
+            }
+            else if (isEmptyCollection(definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(false);
+                offsets.add(offset);
+            }
+            else if (isNotEmptyCollection(definitionLevels.get(i), maxDefinitionLevel)) {
+                collectionIsNull.add(false);
+                offset += countElementsOfNotEmptyCollection(repetitionLevels, maxRepetitionLevelOfCollectionElement, i + 1);
+                offsets.add(offset);
+            }
```

Comment:
`else`?

Reply:
@nezihyigitbasi 
The same logic as described above for undefined struct values applies for collections(arrays/maps).
The only difference is that collections have repeated additional repeated field, that's why they have an additional attribute `offsets`.

For example, column can be a struct "contact" with two fields: string "name" and array "phone"
Hive parquet writer will write such type using the following schema:
```
optional group contact {
    optional binary name (UTF8);
	optional group phone (LIST) {
	    repeated group bag {
	      optional binary array_element (UTF8);
	    }
	}
}
```
Each of those fields will be stored in sepatate Presto block of a corresponding type.
There will be a `RowBlock` for `contact` with two fieldBlocks: a `VariableWidthBlock` for `name` and an `ArrayBlock` for `phones`. `ArrayBlock` will have a nested `VariableWidthBlock` with `array_element` values, along with two helper 
arrays: `offsets` and `collectionIsNull`.

The logic behind populating `offsets` and `collectionIsNull` arrays is the following:

`contact == null` -> array is not defined -> not writing anything to `offsets` and `collectionIsNull`
`phones == null`  -> array is null -> offset from previous array remains unchanged, collectionIsNull = true
`bag == null` -> array is empty (visualy it can be also interpreted as zero bag elements, but in parquet definitition level it is writen as bag is null) -> offset from previous array remains unchanged, collectionIsNull = false
`bag != null` -> array has one or more elements. As `bag` is a `repeated` group, the number of elements can be figured out by looking at it's repetition level. (The most tricky part in counting elements, is the fact, that array's elements can be not primitive. That's why need to count only elements, which belong exactly to this array, not all inner elements of it's nested collections/structs)
            -> offset += array size, collectionIsNull = false